### DISCUSSION TITLE
is it safe to put DATABASE_URL in .evn.production?

### DISCUSSION THREAD

**Discussion by rezarahem** [prisma/prisma#22939] [Q&A]
### Question

Hi 
I have a nextjs app and I want to use generateStaticParams function in one of my routes. But when I deploy the app in the build step, it errors out. 

This is the error
```
Invalid prisma.course.findMany() invocation:

error: Environment variable not found: DATABASE_URL.

--> schema.prisma:7

6 | provider = "postgresql"

7 | url = env("DATABASE_URL")
```

So I figure out this means DATABASE_URL variable in my .evn file is not available during build time or myber my entire .env file itself isn't available. After some googling I found that I could make another env file called .env.production and put the DATABASE_URL  in it. This acutually worked fine, But this made some other problems too. The prisma [docs](https://www.prisma.io/docs/orm/more/development-environment/environment-variables/using-multiple-env-files) says we shouldn't use .evn.production. it seems it's not secure.
So what sould I do? I really want to use generateStaticParams function. it makse the route SSR. 

the is my  use generateStaticParams fn

```js
export async function generateStaticParams() {
  const courses = await prismadb.course.findMany();
  return courses.map(({ courseAddressName }) => ({ courseAddressName }));
}
```

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by bamidele2161** [comment]
Hi @rezarahem,

instead of creating .env.production, 
during the production build process, you can add a script to copy your env file to the build folder.

I hope this helps.

**Reply by rezarahem** [reply]
So what would be the difference. I've already got an env file, .env.production. I want to get rid of them completely. would you explain more? How whould that help?

**Comment by bamidele2161** [comment]
get rid of both .env file and .env.production file?

you need .env file to use prisma, you can remove .env.production file.

**Comment by ludralph** [comment]
Hi @rezarahem 👋 

Thank you for raising this queston.

Which platform are you using to deploy your application? It's common to set environment variables directly in your hosting platform such as vercel, railway, heroku etc. Next.js provides a way to expose environment variables during the build process using `next.config.js`. You can use the `dotenv` package to load environment variables from .env files during development, but in production, you should use server environment variables provided by your hosting platform.

Here's an example of how you can configure Next.js to use environment variables during the build process:
```js
// next.config.js

module.exports = {
  env: {
    DATABASE_URL: process.env.DATABASE_URL,
    // Add other environment variables as needed
  },
};
```
Ensure that `DATABASE_URL` is set as an environment variable in your production environment. This might involve configuring your hosting platform or deployment process to provide this variable.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by rezarahem** [reply]
thanks for reply. I'll test this and share the result.

**Reply by rezarahem** [reply]
I checked this method and still the same result. I get the error that DATABASE_URL is not availble, even hard coded in the config file. I use a service called Liara. I can't really use vercel becuase of sanction on my country Iran. still Liara is very cool. It has the production environment varible you metioned. I set thoes and they work fine in runtime. the thing is during build when I'm deploying I get this error. I get this only with generateStaticParams and also the sitemap file. Meaning if I dont use these no error would happen. Adding .evn.production actully solved the problem. But it seems that's not a safe way.

**Comment by rezarahem** [comment]
I test some more ideas and most of the didn't work. But before that let me tell you something that I missed. I use prisma DATABASE_URI for connection. I didn't know there are two kind of them. One PUBLIC and one PRIVATE. In my production environment varible at the hosting platform (which for me is Liara, something like vercel) I set DATABASE_URL to that privte url. the thing is you can only use that in production and only on the platform's server, any computer outside of their severs can't access any prod envs. 

So all of my test with private url faild. Finally I used the public one that work with two methed, One making a .env.production file and set url varible inside which prisma doc says you souldn't use. The other way that worked actually is really cool but I don't really know is it safe or not. So for this you can put you DATABASE_URL with its value in the build script in the package.json; like this

```
scripts": {
    "dev": "next dev",
    "prebuild": "npx prisma generate",
    "build": "DATABASE_URL=postgresql:<yout public url>  next build",
    "start": "next start",
    "lint": "next lint"
  }
```

Also you could turn off public url any time. But I just learn this now. I need to test more and see if it would work intended.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
